File: frontend/kiosk/src/app/model/SensorBoxDTO.ts
export class SensorBoxDTO {
  room: string;
  floor: string;
  co2?: number;
  humidity?: number;
  motion?: number;
  neopixel?: number;
  noise?: number;
  pressure?: number;
  rssi?: number;
  temperature?: number;
  timestamp: number;

  constructor(
    room: string,
    floor: string,
    timestamp: number,
    co2?: number,
    humidity?: number,
    motion?: number,
    neopixel?: number,
    noise?: number,
    pressure?: number,
    rssi?: number,
    temperature?: number
  ) {
    this.room = room;
    this.floor = floor;
    this.timestamp = timestamp;
    this.co2 = co2;
    this.humidity = humidity;
    this.motion = motion;
    this.neopixel = neopixel;
    this.noise = noise;
    this.pressure = pressure;
    this.rssi = rssi;
    this.temperature = temperature;
  }

  toString(): string {
    return `SensorBoxDTO { Room: ${this.room}, Floor: ${this.floor}, Timestamp: ${this.timestamp}, CO2: ${this.co2?.toFixed(
      2
    )}, Humidity: ${this.humidity?.toFixed(2)}, Motion: ${this.motion?.toFixed(
      2
    )}, Noise: ${this.noise?.toFixed(2)}, Pressure: ${this.pressure?.toFixed(
      2
    )}, RSSI: ${this.rssi?.toFixed(2)}, Temperature: ${this.temperature?.toFixed(
      2
    )} }`;
  }
}

File: backend/leoenergy/src/main/java/at/htl/leoenergy/entity/SensorBoxDTO.java
package at.htl.leoenergy.entity;

public class SensorBoxDTO {
    private String room;
    private String floor;
    private Double co2;
    private Double humidity;
    private Double motion;
    private Double neopixel;
    private Double noise;
    private Double pressure;
    private Double rssi;
    private Double temperature;
    private long timestamp; // Time for the latest reading

    public SensorBoxDTO() {
    }

    public SensorBoxDTO(String room, String floor, long timestamp) {
        this.room = room;
        this.floor = floor;
        this.timestamp = timestamp;
    }

    // Getters and setters for all fields

    public String getRoom() {
        return room;
    }

    public void setRoom(String room) {
        this.room = room;
    }

    public String getFloor() {
        return floor;
    }

    public void setFloor(String floor) {
        this.floor = floor;
    }

    public Double getCo2() {
        return co2;
    }

    public void setCo2(Double co2) {
        this.co2 = co2;
    }

    public Double getHumidity() {
        return humidity;
    }

    public void setHumidity(Double humidity) {
        this.humidity = humidity;
    }

    public Double getMotion() {
        return motion;
    }

    public void setMotion(Double motion) {
        this.motion = motion;
    }

    public Double getNeopixel() {
        return neopixel;
    }

    public void setNeopixel(Double neopixel) {
        this.neopixel = neopixel;
    }

    public Double getNoise() {
        return noise;
    }

    public void setNoise(Double noise) {
        this.noise = noise;
    }

    public Double getPressure() {
        return pressure;
    }

    public void setPressure(Double pressure) {
        this.pressure = pressure;
    }

    public Double getRssi() {
        return rssi;
    }

    public void setRssi(Double rssi) {
        this.rssi = rssi;
    }

    public Double getTemperature() {
        return temperature;
    }

    public void setTemperature(Double temperature) {
        this.temperature = temperature;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    @Override
    public String toString() {
        return String.format(
                "SensorBoxDTO { Room: %s, Floor: %s, Timestamp: %d, CO2: %.2f, Humidity: %.2f, Motion: %.2f, Noise: %.2f, Pressure: %.2f, RSSI: %.2f, Temperature: %.2f }",
                room, floor, timestamp, co2, humidity, motion, noise, pressure, rssi, temperature
        );
    }

    public static SensorBoxDTO deleteFaultyValues(SensorBoxDTO dto) {
        dto.co2 = dto.co2 >= 0 && dto.co2 <= 1_000_000 ? dto.co2 : Double.NaN;
        dto.humidity = dto.humidity >= 0 && dto.humidity <= 100 ? dto.humidity : Double.NaN;
        dto.noise = dto.noise >= 0 && dto.noise <= 194 ? dto.noise : Double.NaN;
        dto.pressure = dto.pressure >= 0 ? dto.pressure : Double.NaN;
        dto.temperature = dto.temperature >= -273.15 ? dto.temperature : Double.NaN;
        return dto;
    }
}
